p2p: opportunistically accept 1-parent-1-child packages

glozow opened the pull request:
This enables 1p1c packages to propagate in the "happy case" (i.e. not reliable if there are adversaries) and contains a lot of package relay-related code. See https://github.com/bitcoin/bitcoin/issues/27463 for overall package relay tracking.

Rationale: This is "non-robust 1-parent-1-child package relay" which is immediately useful.
- Relaying 1-parent-1-child CPFP when mempool min feerate is high would be a subset of all package relay use cases, but a pretty significant improvement over what we have today, where such transactions don't propagate at all. [1]
- Today, a miner can run this with a normal/small maxmempool to get revenue from 1p1c CPFP'd transactions without losing out on the ones with parents below mempool minimum feerate.
- The majority of this code is useful for building more featureful/robust package relay e.g. see the code in #27742.

The first 2 commits are followups from #29619:
- https://github.com/bitcoin/bitcoin/pull/29619#discussion_r1523094034
- https://github.com/bitcoin/bitcoin/pull/29619#discussion_r1519819257

Q: What makes this short of a more full package relay feature?

(1) it only supports packages in which 1 of the parents needs to be CPFP'd by the child. That includes 1-parent-1-child packages and situations in which the other parents already pay for themselves (and are thus in mempool already when the package is submitted). More general package relay is a future improvement that requires more engineering in mempool and validation - see #27463.

(2) We rely on having kept the child in orphanage, and don't make any attempt to protect it while we wait to receive the parent. If we are experiencing a lot of orphanage churn (e.g. an adversary is purposefully sending us a lot of transactions with missing inputs), we will fail to submit packages. This limitation has been around for 12+ years, see #27742 which adds a token bucket scheme for protecting package-related orphans at a limited rate per peer.

(3) Our orphan-handling logic is somewhat opportunistic; we don't make much effort to resolve an orphan beyond asking the child's sender for the parents. This means we may miss packages if the first sender fails to give us the parent (intentionally or unintentionally). To make this more robust, we need receiver-side logic to retry orphan resolution with multiple peers. This is also an existing problem which has a proposed solution in #28031.

[1]: see this writeup and its links https://github.com/bitcoin/bips/blob/02ec218c7857ef60914e9a3d383b68caf987f70b/bip-0331.mediawiki#propagate-high-feerate-transactions

glozow committed:
[functional test] opportunistic 1p1c package submission

glozow committed:
[p2p] add separate rejections cache for reconsiderable txns

glozow commented:
Rebased and fixed CI. This is in draft because I'm focusing on v3 stuff, can be reviewed for its approach.

instagibbs commented:
ready for un-draft?

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3036,6 +3045,112 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
+    // Maybe punish peer that gave us an tx
+    MaybePunishNodeForTx(nodeid, state);
+
+    switch (state.GetResult()) {
```
we used to call `m_orphanage.EraseTx(orphanHash);` for each non-`TX_MISSING_INPUTS` possibility, but due to early return we don't. Can we sync this back up or justify each case this has changed for?

glozow replied:
Old behavior should be preserved now

instagibbs commented on src/net_processing.cpp:
with changes, we now add to extra txns even if witness stripped; seems wrong

glozow replied:
fixed

instagibbs commented on test/functional/p2p_1p1c_package_relay.py:
```diff
@@ -0,0 +1,223 @@
+        self.log.info("Check end-to-end package relay across multiple nodes")
+        self.packages_to_submit = []
+        self.transactions_to_presend = [[]] * self.num_nodes
+        self.replacement_packages = []
```
this is never set

glozow replied:
Oops, forgot to delete from a package RBF test I built on top. Fixed

instagibbs reviewed (commented):
I focused on the first two refactoring commits. Might be helpful zeroing in on these since we probably need to justify each small behavior change or make sure we're aligning with old behavior.

brunoerg commented on test/functional/p2p_1p1c_package_relay.py:
```diff
@@ -0,0 +1,223 @@
+class PackageRelayTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 4
```
In 19015e3c9aed5709e776ef15bf9e73f126c7ea29: It seems to be using 3 nodes instead of 4. Or am I missing something?

E.g.:

```diff
diff --git a/test/functional/p2p_1p1c_package_relay.py b/test/functional/p2p_1p1c_package_relay.py
index 6a3ba7c3df..dda671dede 100755
--- a/test/functional/p2p_1p1c_package_relay.py
+++ b/test/functional/p2p_1p1c_package_relay.py
@@ -36,7 +36,7 @@ FEERATE_1SAT_VB = Decimal("0.00001000")
 class PackageRelayTest(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = True
-        self.num_nodes = 4
+        self.num_nodes = 3
         self.extra_args = [[
             "-datacarriersize=100000",
             "-maxmempool=5",
@@ -85,7 +85,7 @@ class PackageRelayTest(BitcoinTestFramework):
         # Child should already be in orphanage
         self.transactions_to_presend[1] = [high_fee_child["tx"]]
         # Parent would have been previously rejected
-        self.transactions_to_presend[3] = [low_fee_parent["tx"]]
+        self.transactions_to_presend[2] = [low_fee_parent["tx"]]
 
     def test_individual_logic(self):
         node = self.nodes[0]
@@ -213,7 +213,7 @@ class PackageRelayTest(BitcoinTestFramework):
 
         self.log.info("Submit replacement package to node3")
         for package_hex in self.replacement_packages:
-            self.nodes[3].submitpackage(package_hex)
+            self.nodes[2].submitpackage(package_hex)
 
         self.log.info("Wait for mempools to sync")
         self.sync_mempools(timeout=20)
```

glozow replied:
The idea is this is a "network" test to see that the package propagates across multiple hops. 1 node pre-receives the parent and 1 node pre-receives the child, and there is a node in between.

brunoerg replied:
Got it, thank you.

glozow commented:
Working on pulling the first 2 commits out into a separate PR.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3036,6 +3045,112 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
+        // adding such txids to the reject filter would potentially
+        // interfere with relay of valid transactions from peers that
+        // do not support wtxid-based relay. See
+        // https://github.com/bitcoin/bitcoin/issues/8279 for details.
```
@sdaftuar honestly this comment/link doesn't make sense to me. If it's witness-stripped, we can't add it to `m_recent_rejects` since that would blind us to any real version of the tx?

sdaftuar replied:
I had to remind myself of how this works, but the concern contemplated in that comment is that if the **only** copy of a non-witness-stripped version of a transaction is from a node that has not yet upgraded to wtxid-based-relay, then an adversary could blind us to that transaction by relaying a witness-stripped version of it (and then, since txid==wtxid for witness-stripped transactions, the hash that we'd use to fetch the transaction from the non-wtxid-relay peer who has the correct version would be in our `m_recent_rejects`, and we wouldn't ever request it).

At the time that I wrote that comment, I think I also said that once wtxid-based-relay is sufficiently deployed, that we could stop worrying about this...  I don't recall when wtxid-relay was deployed but maybe we're at that point already?

There's another link to a comment thread in the PR that implemented wtxid-relay which discusses the potential impact on downstream projects, so maybe if we were to make a behavior change we should communicate that in advance so that no one is surprised.

glozow replied:
It seems like the comment thread is concerned with what happens if:
(1) the sender is using non-wtxidrelay (and thus only announces transactions by txid),
(2) all of its peers are caching witness-stripped rejections by wtxid (i.e. txid),
(3) there is an attacker trying to censor the sender's tx by sending witness-stripped versions of the tx ahead of the sender.

This isn't a problem if (1) isn't true, i.e. the sender is using wtxidrelay, e.g. any non-eol version of Bitcoin Core. If we see that the vast majority of nodes (that send transactions) are sending a `wtxidrelay` message, perhaps it would be safe to remove the special casing (with a message to downstream projects)?

glozow replied:
Resolving as out of scope for this PR. Also fwiw, btcd doing announcements by txid seems to be a reason to keep it for now...

glozow committed:
[p2p] opportunistically accept 1-parent-1-child packages

glozow commented on src/net_processing.cpp:
```diff
@@ -3078,54 +3193,15 @@ bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
-                    m_recent_rejects.insert(porphanTx->GetHash().ToUint256());
-                }
-            }
-            m_orphanage.EraseTx(orphanHash);
```
Rewrote to be more strict refactor to give reviewers less to think about. I do think it makes sense that if a parent has been rejected since the last time we looked at this orphan, we get rid of it, but I suppose we can think about that later.

glozow commented:
Opened #29619 for the first 2 commits, moving this to draft.

glozow committed:
guard against MempoolAcceptResult::m_replaced_transactions

It should never be a nullopt when the transaction result is valid -
Assume() this is the case. However, as a belt-and-suspenders just in
case it is nullopt, use an empty list.

glozow committed:
[doc] restore comment about why we check if ptx HasWitness before caching rejected txid

glozow commented:
rebased and added followups from #29619

instagibbs reviewed (commented):
some comments, running patch on a listening node longer-term to test behavior

glozow commented on src/net_processing.cpp:
```diff
@@ -846,6 +846,23 @@ class PeerManagerImpl final : public PeerManager
+     *
+     * Reset this filter when the chain tip changes.
+     *
+     * Parameters are picked to be the same false positive rate but half the capacity as
```
Not much science, no. It makes sense to me to use the same false positive rate. As for size, I'll run my node for a bit longer and maybe use the ratio of `m_recent_rejects` / `m_recent_rejects_reconsiderable` usage to give a more scientific number.

glozow commented on src/net_processing.cpp:
```diff
@@ -806,7 +806,7 @@ class PeerManagerImpl final : public PeerManager
     std::atomic<std::chrono::seconds> m_block_stalling_timeout{BLOCK_STALLING_TIMEOUT_DEFAULT};
 
-    bool AlreadyHaveTx(const GenTxid& gtxid)
+    bool AlreadyHaveTx(const GenTxid& gtxid, bool include_reconsiderable)
```
Vestigial from a previous iteration of this PR and future changes. Will drop

instagibbs replied:
can confirm, changes seem logical, and there's test coverage for each `AlreadyHaveTx` instance

glozow commented on src/net_processing.cpp:
```diff
@@ -2209,6 +2226,8 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid)
 
     if (m_orphanage.HaveTx(gtxid)) return true;
 
+    if (include_reconsiderable && m_recent_rejects_reconsiderable.contains(gtxid.GetHash())) return true;
```
done

glozow commented on src/net_processing.cpp:
```diff
@@ -3181,6 +3187,92 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    const auto cpfp_candidates{m_orphanage.GetChildren(ptx, nodeid)};
+    bool tried_package_cpfp{false};
+
+    const auto& txid{ptx->GetHash()};
```
done

glozow commented on test/functional/p2p_1p1c_package_relay.py:
```diff
@@ -0,0 +1,222 @@
+            # out the in-flight requests.  It is currently required for the test to pass right now,
+            # because the node will not (re)try requesting orphan parents from multiple peers if the
+            # first one fails.
+            # TODO: remove this and test that the node retries orphan resolution with other peers
```
I guess not a lot. But it is worth checking that it having seen (and rejected) the child before shouldn't impact its acceptance of the package later?

glozow commented:
Rebased for silent conflict with #28950

glozow commented on test/functional/p2p_1p1c_package_relay.py:
```diff
@@ -0,0 +1,222 @@
+        self.log.info("Check 1p1c validation logic on a single node")
+        self.test_individual_logic()
+
+        self.log.info("Check end-to-end package relay across multiple nodes")
```
done

dergoegge commented:
> Enable 1-parent-1-child package relay 🚀 (just using the existing protocol messages).

Imo, this isn't really what this PR does because "package relay" (at least to me) is something that is robust and users can rely on. As you note yourself, there is more work required to make this robust. 

> To make this more robust, we need receiver-side logic to retry orphan resolution with multiple peers.
> To make this robust, we need to make efforts to protect some amount of orphans from eviction.

Why are we not making things more robust in or before this PR?

instagibbs commented:
> Why are we not making things more robust in or before this PR?

Why would we tie up an improvement on relay on other improvements to relay? This feature is useful by itself in non-adversarial network conditions, which ends up being deliverable security and a bit of increased miner fee revenue.

Of course, if we have a orphanage hardening PR ready for merge, we should merge that.

glozow commented:
> Why are we not making things more robust in or before this PR?

Why have we been okay with handling any transactions with missing parents despite all of these orphanage problems, for more than a decade? Clearly because it's useful even if not 100% reliable. There are various improvements we can make to relay. I have already implemented them, and decided I would prefer to propose the more impactful improvements before the less impactful ones.

dergoegge commented:
> that sounds like confusing semantic games to me? We can add "non-robust" to the title to be clearer if we think it can get merged...

I think my issue is with the PR description and how this will be marketed, not the title. The title describes it quite well: we will sometimes accept 1p1c packages (although it could also mention that these packages are constructed from the orphanage). That is different from package relay (again, to me "package relay" implies a level of reliability).

> This feature is useful by itself in non-adversarial network conditions, which ends up being deliverable security

"deliverable security": Is this referring to the increased cost of having to mess with relay to prevent pkgs from relaying? as opposed to the current state where these pkgs just don't relay at all?

> Why have we been okay with handling any transactions with missing parents despite all of these orphanage problems, for more than a decade?

Because it saves bandwidth in the honest case. Afaict, relay would be reliable without the orphanage, although much more bandwidth intensive.

---

I guess my wording was confusing? I have no problem with the approach in this PR nor was I suggesting to block this on other relay improvements. I was just commenting on the PR description and asking a question.

There clearly are good answers to my question but I don't appreciate the condescending tone after being asked to review this PR.

instagibbs commented:
> "deliverable security": Is this referring to the increased cost of having to mess with relay to prevent pkgs from relaying? as opposed to the current state where these pkgs just don't relay at all?

Assuming there's a path of nodes to a miner in which they are not having their orphanages aggressively churned by an adversary,  these size 2 packages should be able to propagate. So if say a LN node goes offline or refuses to sign an updated commitment transaction that raises the transaction fee, minfee rises, the counterparty can still go to chain. It raises the bar a bit for an adversary, and the more benign cases can be resolved in the average case, letting people deploy liquidity elsewhere.

FWIW I ran this branch over the last week, and when my minfee wasn't 1 I was getting about one extra package relayed every half hour.

glozow commented:
I've updated the PR description to be more descriptive about what this PR does and doesn't do. The 1p1c "package relay" thing wasn't meant as false advertising; I added it to the PR description 1 month after opening because it seemed like people were underestimating how useful this is (someone was surprised when I said this PR propagates 1p1c packages).

I apologize for the tone - the comment read to me like a blocking criticism since it wasn't accompanied by any other review comments. Hopefully the suggestion has been adequately addressed with the description update now.

Roasbeef commented:
Popping in here to mention that even w/ the limitations described above, this would be super useful for LN as is. Today we're able to use the anchor outputs on commitment transactions to CPFP a force close (* caveats re pinning, RBF, etc, etc). This works OK during calm mempool conditions, but if a rapid spike occurs that causes the bottom of the mempool to fall out, then CPFP no longer works as the parent has been evicted. During one of the recent more persistent fee spikes, we saw this happening rather frequently, which then led to cascading force closes (outgoing HTLC can't timeout due to non confirmation, incoming times out, repeat) and a lots of pain and user confusion (can only wait or use a transaction "accelerator"). 

IIUC, with this PR, upgraded peers will optimistically attempt to fetch the parent (the commitment txn) when they see a transaction that appears to be fee bumping it (anchor output spend). With this behavior, on a best effort basis, if we had another massive spike, then commitment transactions would be able to confirm as upgraded peers would fetch the parent to consider/propagate. I think this is very attractive as it doesn't necessarily require a new p2p upgrade to gain the benefits, instead as nodes start to progressively update, then more of these otherwise unrecognized packages will start to propagate. If major LN node operators are made aware of this patch, and start to run it with the node backing their LN nodes, then assuming a relay path to miners (who seem to be willing to mine/relay just about anything these days), this would be a rather nice stop gap while the greater v3/cluster work proceeds in the background.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4420,7 +4532,10 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-                    if (!AlreadyHaveTx(gtxid, /*include_reconsiderable=*/true)) AddTxAnnouncement(pfrom, gtxid, current_time);
+                    // Skip m_recent_rejects_reconsiderable because the missing parent may have been
+                    // previously rejected for being too low feerate, and this orphan may be able to
+                    // CPFP it if we consider them as a package.
```
this change also only matters for the non-segwit case since we're checking via txid, correct?

glozow replied:
Correct. If it's a witness tx, we would have cached the error by wtxid but the request is by txid, so this would miss it.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -4378,6 +4482,14 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             // peer simply for relaying a tx that our m_recent_rejects has caught,
             // regardless of false positives.
             return;
+        } else if (m_recent_rejects_reconsiderable.contains(wtxid)) {
```
tiny preference for putting this case first

instagibbs reviewed (commented):
no issues running the patch so far, some more comments

brunoerg commented on src/txorphanage.cpp:
```diff
@@ -241,3 +241,19 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
     }
 }
+
+std::vector<CTransactionRef> TxOrphanage::GetChildren(const CTransactionRef& parent, NodeId peer) const
```
In f4d8fe713a036c4d2c1e7c2329077c34f75c8608: It seems `NodeId peer` is not used in `GetChildren`.

glozow committed:
[txorphanage] add method to get all orphans spending a tx

glozow committed:
[unit test] TxOrphanage::GetChildrenFrom*

glozow committed:
[fuzz] break out parent functions and add GetChildrenFrom* coverage

It's very hard to randomly construct a transaction that would be the
parent of an existing orphanage tx. For functions like
AddChildrenToWorkSet and GetChildren that take orphan parents, use a tx
that was previously constructed.

glozow commented on test/functional/p2p_1p1c_package_relay.py:
```diff
@@ -0,0 +1,232 @@
+        self.extra_args = [[
+            "-datacarriersize=100000",
+            "-maxmempool=5",
+            "-whitelist=noban@127.0.0.1",  # immediate tx relay
```
done

glozow commented on test/functional/p2p_1p1c_package_relay.py:
```diff
@@ -0,0 +1,232 @@
+            assert_equal(node.getmempoolinfo()['minrelaytxfee'], FEERATE_1SAT_VB)
+            assert_greater_than(node.getmempoolinfo()['mempoolminfee'], FEERATE_1SAT_VB)
+
+    def create_packages(self):
```
not sure if I achieved spooky, but I'm having it return the lists now :ghost:

glozow commented on src/txorphanage.h:
```diff
@@ -51,6 +51,10 @@ class TxOrphanage {
     bool HaveTxToReconsider(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);;
 
+    /** Get all children of this parent. */
+    std::vector<CTransactionRef> GetChildren(const CTransactionRef& parent, NodeId peer) const
```
added unit tests and fuzz coverage

glozow commented:
Thanks for the review, addressed comments and
- fixed a bug in `GetChildren`
- expanded + added more docs to the individual logic tests
- got rid of the `assert_debug_log` lines in the functional tests. The only "coverage" that's lost is us skipping children with which we've already tried+failed the package submission. I have a [branch with them](https://github.com/glozow/bitcoin/blob/2024-03-28970-test-logs/test/functional/p2p_opportunistic_1p1c.py) if you want to use it to test the code, but log asserts didn't feel appropriate to have in the functional test.
- added more logging, realized it was annoyingly hard to look up success/fail package evals while searching logs on my node

fwiw running this on a `-maxmempool=150` node, for the past few days, on average:
- 222 attempted package validations per day
- 85 packages (so 170 txns) accepted per day
- 57% of the package transactions accepted ended up in a block. These are all transactions that we would have otherwise rejected, so this logic is definitely useful at helping us maximizing the fees in our limited mempool space.

theStack reviewed (commented):
Concept ACK

Found some nitty stuff mostly in tests so far, planning to do another review round next week.

murchandamus reviewed (commented):
Did a quick pass. I am surprised we are able to get this big of an improvement without p2p changes, seems like a big win. Big Concept ACK, but I must admit this part of the codebase is a bit out of my wheelhouse.

glozow commented on src/net_processing.cpp:
```diff
@@ -4354,7 +4354,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         const TxValidationState& state = result.m_state;
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
-            ProcessValidTx(pfrom.GetId(), ptx, result.m_replaced_transactions.value());
+            Assume(result.m_replaced_transactions.has_value());
+            std::list<CTransactionRef> empty_replacement_list;
+            ProcessValidTx(pfrom.GetId(), ptx, result.m_replaced_transactions.value_or(empty_replacement_list));
```
> it is unclear to me from the commit message and the code change how this change fits in the context.

(Note that this commit is a followup from #29619)

Yes, it should always have a value when the result is VALID. This is just adding a belt-and-suspenders juuust in case.

murchandamus replied:
Thanks! Great, I’ll attempt another more thorough review at a later time.

glozow replied:
Elaborated in commit message

instagibbs commented on src/net_processing.cpp:
```diff
@@ -6061,7 +6186,9 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
         }
         for (const GenTxid& gtxid : requestable) {
-            if (!AlreadyHaveTx(gtxid, /*include_reconsiderable=*/true)) {
+            // Skip m_recent_rejects_reconsiderable because we may be requesting a missing parent
```
I have to admit I'm struggling on these "skip" comments and cases. Every time I read this I have trouble re-deriving the logic.

instagibbs replied:
think you forgot to push?

glozow replied:
> think you forgot to push?

Github published everything instead of adding my comment to the group -_- yes, pushing in a second.

instagibbs reviewed (commented):
reviewed through 81912ae6649fac5a0d671d2cabe344f787193997

LGTM so far

will review tests

instagibbs commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,203 @@
+
+        node.disconnect_p2ps()
+
+    def run_test(self):
```
add a case where:
1) CONSENSUS-invalid child is propagated to peer
2) low_fee parent is given
3) package evaluation is attempted
4) peer is disconnected

to ensure we're handling invalid tx properly

glozow replied:
Wait oof, we don't require the orphanage child to be provided by the same peer who sent the low-feerate tx. I think this would mean you can get other people disconnected by sending a bogus child of the package they're sending.

instagibbs replied:
looks like peer id needs to be used again: https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1542964936

glozow replied:
Hm no, I think we should return pairs of tx and the `fromPeer`, and just attribute the error to the right peer. If we restrict it to same sender, then we can also block packages by sending the children fast and refusing to send the parent.

glozow replied:
Last push should fix this with the above approach. Still need to write a test case for the peers being different.

instagibbs commented:
reviewed through c5e196b8e7b610cc9e6321b76fee4f0c45c1448e

glozow commented on src/net_processing.cpp:
```diff
@@ -806,7 +806,7 @@ class PeerManagerImpl final : public PeerManager
     std::atomic<std::chrono::seconds> m_block_stalling_timeout{BLOCK_STALLING_TIMEOUT_DEFAULT};
 
-    bool AlreadyHaveTx(const GenTxid& gtxid)
+    bool AlreadyHaveTx(const GenTxid& gtxid, bool include_reconsiderable)
```
> we should have tests(in master?) for this I hope.

apparently not?

glozow replied:
marking as resolved since #29827 does this

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3153,6 +3208,103 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+
+    // Iterate backwards to erase in-package descendants from the orphanage before they become
+    // relevant in AddChildrenToWorkSet.
+    for (auto i{package.size() - 1}; i >= 0; --i) {
```
```
net_processing.cpp:3224:40: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
 3224 |     for (auto i{package.size() - 1}; i >= 0; --i) {
      |            
```

instagibbs replied:
this is also causing the sanitizer run failure

glozow replied:
Woops yes will fix

glozow commented:
Rebased for #29735 and fixed failure.

glozow commented:
Update since my last comment https://github.com/bitcoin/bitcoin/pull/28970#issuecomment-2025812846
Running for 20 days:
- My node made 5218 package evaluation attempts (x2 = 10,436 transactions). That's ~260 attempts per day.
- Of those attempts, 2260 were successful (x2 = 4520 transactions). That's ~43% acceptance rate.
- Some were repeats. I have 4202 unique txids accepted through package evaluation. I don't know if they were repeated packages or not.
- Of the accepted transactions, 3232 (77%) were later mined in a block.

mzumsande reviewed (commented):
Started reviewing - haven't looked at the tests yet.

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,108 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    NodeId orphan_sender{-1};
+    Assume(m_recent_rejects_reconsiderable.contains(parent_wtxid.ToUint256()));
+
+    // Create a 1p1c package using the first child that hasn't already been rejected. Sort
```
Yes exactly, this is to bound computation (imagine if somebody sent us 100 fake orphans descended from 1 transaction and we processed them all here). My idea was to do something similar to regular orphan processing, where we have a work queue and limit to 1 item per `ProcessMessages`.

There is no work queue here, though, and we drop the parent as soon as we try 1 (pass or fail). At coredev, we discussed adding a work queue for 1p1c as well. However, since it involves finding a way to store the low feerate parent, and we have plenty of low hanging fruit for improving orphan handling, we'll save that for a later PR.

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,108 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+        // If we already tried a package and failed for any reason, the combined hash was
+        // cached in m_recent_rejects_reconsiderable.
+        Package maybe_cpfp_package{ptx, cpfp_candidates.at(index).first};
+        if (!m_recent_rejects_reconsiderable.contains(GetPackageHash(maybe_cpfp_package))) {
```
Yes I think we could use either `m_recent_rejects_reconsiderable` and `m_recent_rejects` right now to get the same behavior.

I suppose one mild benefit of using `m_recent_rejects_reconsiderable` is that our `m_recent_rejects` bloom filter churns less frequently.

The other benefit is extensibility in the future. In more general ancestor package relay, we could reject a parent+child for being too low feerate, but later accept it as parent+child+grandchild (where the grandchild is very high feerate).

instagibbs replied:
> The other benefit is extensibility in the future. In more general ancestor package relay, we could reject a parent+child for being too low feerate, but later accept it as parent+child+grandchild (where the grandchild is very high feerate).

Perhaps this doesn't matter but I'm not sure I understand the distinction here. We need the combined hash committed *somewhere* in a bloom filter to not fetch the same ancestor package again. If it's different at all, we'll fetch it regardless of which filter we add it to?

glozow replied:
Hm I think you're right, it wouldn't make a difference with downloads. Crossing that part out. Were we thinking of this within validation maybe? Linearize + chunk the package, see that a chunk has already been rejected as too low feerate, drop it?

glozow committed:
[txpackages] add canonical way to get hash of package

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,108 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    AssertLockHeld(g_msgproc_mutex);
+    AssertLockHeld(cs_main);
+
+    const auto cpfp_candidates{m_orphanage.GetChildren(ptx)};
```
Done, and edited the logging a bit

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,108 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    if (package_result.m_state.IsInvalid()) {
+        m_recent_rejects_reconsiderable.insert(GetPackageHash(package));
+    }
+    if (!Assume(!package.empty())) return;
```
added

glozow commented on src/net_processing.cpp:
```diff
@@ -3097,7 +3130,14 @@ void PeerManagerImpl::ProcessInvalidTx(NodeId nodeid, const CTransactionRef& ptx
+            // If the result is TX_RECONSIDERABLE, add it to m_recent_rejects_reconsiderable
+            // because we should not download or submit this transaction by itself again, but may
+            // submit it as part of a package later.
+            m_recent_rejects_reconsiderable.insert(ptx->GetWitnessHash().ToUint256());
```
updated

glozow reviewed (commented):
thanks for the review @mzumsande!

instagibbs commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,241 @@
+        parent_txid_int = int(low_fee_parent["txid"], 16)
+        peer_sender.wait_for_getdata([parent_txid_int])
+
+        # 3. Sender relays the parent. Parent+Child are evaluated as a package and accepted.
```
```Suggestion
        # 3. Sender relays the parent. Parent+Child are evaluated as a package and rejected.
```

instagibbs commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,241 @@
+        tx_orphan_bad_wit.wit.vtxinwit.append(CTxInWitness())
+        tx_orphan_bad_wit.wit.vtxinwit[0].scriptWitness.stack = [b'garbage']
+
+        peer_sender = node.add_p2p_connection(P2PInterface())
```
would be good to explicitly test that the parent-giver isn't punished in this scenario, and test if parent is consensus-bad it results in something expected.

```
diff --git a/test/functional/p2p_opportunistic_1p1c.py b/test/functional/p2p_opportunistic_1p1c.py
index 603cbf08a9..1c887289dc 100755
--- a/test/functional/p2p_opportunistic_1p1c.py
+++ b/test/functional/p2p_opportunistic_1p1c.py
@@ -164,78 +164,129 @@ class PackageRelayTest(BitcoinTestFramework):
         # itself and with a child. This is necessary, otherwise high_fee_child can be censored.
         parent_txid_int = int(low_fee_parent["txid"], 16)
         peer_sender.wait_for_getdata([parent_txid_int])
 
         # 7. The low feerate parent + high feerate child are submitted as a package.
         peer_sender.send_and_ping(msg_tx(low_fee_parent["tx"]))
 
         # 8. Both transactions should now be in mempool
         node_mempool = node.getrawmempool()
         assert low_fee_parent["txid"] in node_mempool
         assert high_fee_child["txid"] in node_mempool
 
         node.disconnect_p2ps()
 
     def test_orphan_consensus_failure(self):
         node = self.nodes[0]
         low_fee_parent = self.wallet.create_self_transfer(fee_rate=FEERATE_1SAT_VB, confirmed_only=True)
         coin = low_fee_parent["new_utxo"]
         address = node.get_deterministic_priv_key().address
         # Create raw transaction spending the parent, but with no signature (a consensus error).
         hex_orphan_no_sig = node.createrawtransaction([{"txid": coin["txid"], "vout": coin["vout"]}], {address : coin["value"] - Decimal("0.0001")})
         tx_orphan_bad_wit = tx_from_hex(hex_orphan_no_sig)
         tx_orphan_bad_wit.wit.vtxinwit.append(CTxInWitness())
         tx_orphan_bad_wit.wit.vtxinwit[0].scriptWitness.stack = [b'garbage']
 
-        peer_sender = node.add_p2p_connection(P2PInterface())
+        child_peer_sender = node.add_p2p_connection(P2PInterface())
+        parent_peer_sender = node.add_p2p_connection(P2PInterface())
 
         # 1. Child is received first. It is missing an input.
         child_wtxid_int = int(tx_orphan_bad_wit.getwtxid(), 16)
-        peer_sender.send_and_ping(msg_inv([CInv(t=MSG_WTX, h=child_wtxid_int)]))
-        peer_sender.wait_for_getdata([child_wtxid_int])
-        peer_sender.send_and_ping(msg_tx(tx_orphan_bad_wit))
+        child_peer_sender.send_and_ping(msg_inv([CInv(t=MSG_WTX, h=child_wtxid_int)]))
+        child_peer_sender.wait_for_getdata([child_wtxid_int])
+        child_peer_sender.send_and_ping(msg_tx(tx_orphan_bad_wit))
 
         # 2. Node requests the missing parent by txid.
         parent_txid_int = int(low_fee_parent["txid"], 16)
-        peer_sender.wait_for_getdata([parent_txid_int])
+        child_peer_sender.wait_for_getdata([parent_txid_int])
 
-        # 3. Sender relays the parent. Parent+Child are evaluated as a package and accepted.
-        peer_sender.send_message(msg_tx(low_fee_parent["tx"]))
+        # 3. "Honest" sender relays the parent. Parent+Child are evaluated as a package and accepted.
+        parent_peer_sender.send_message(msg_tx(low_fee_parent["tx"]))
 
         # 4. Transactions should not be in mempool.
         node_mempool = node.getrawmempool()
         assert low_fee_parent["txid"] not in node_mempool
         assert tx_orphan_bad_wit.rehash() not in node_mempool
 
         # 5. Peer sent a consensus-invalid transaction.
-        peer_sender.wait_for_disconnect()
+        child_peer_sender.wait_for_disconnect()
+
+        # 6. "Honest" peer unpunished
+        parent_peer_sender.sync_with_ping()
+
+    def test_parent_consensus_failure(self):
+        node = self.nodes[0]
+        low_fee_parent = self.wallet.create_self_transfer(fee_rate=FEERATE_1SAT_VB, confirmed_only=True)
+
+        # Add bad sigature to parent
+        tx_parent_bad_wit = tx_from_hex(low_fee_parent["hex"])
+        tx_parent_bad_wit.wit.vtxinwit.append(CTxInWitness())
+        tx_parent_bad_wit.wit.vtxinwit[0].scriptWitness.stack = [b'garbage']
+
+        coin = low_fee_parent["new_utxo"]
+        address = node.get_deterministic_priv_key().address
+        # Create raw transaction spending the parent, but with no signature (a consensus error).
+        hex_orphan_no_sig = node.createrawtransaction([{"txid": coin["txid"], "vout": coin["vout"]}], {address : coin["value"] - Decimal("0.0001")})
+        tx_orphan_bad_wit = tx_from_hex(hex_orphan_no_sig)
+        tx_orphan_bad_wit.wit.vtxinwit.append(CTxInWitness())
+        tx_orphan_bad_wit.wit.vtxinwit[0].scriptWitness.stack = [b'garbage']
+
+        child_peer_sender = node.add_p2p_connection(P2PInterface())
+        parent_peer_sender = node.add_p2p_connection(P2PInterface())
+
+        # 1. Child is received first. It is missing an input.
+        child_wtxid_int = int(tx_orphan_bad_wit.getwtxid(), 16)
+        child_peer_sender.send_and_ping(msg_inv([CInv(t=MSG_WTX, h=child_wtxid_int)]))
+        child_peer_sender.wait_for_getdata([child_wtxid_int])
+        child_peer_sender.send_and_ping(msg_tx(tx_orphan_bad_wit))
+
+        # 2. Node requests the missing parent by txid.
+        parent_txid_int = int(tx_parent_bad_wit.rehash(), 16)
+        child_peer_sender.wait_for_getdata([parent_txid_int])
+
+        # 3. Parent sender relays the parent. Only parent should be evaluated.
+        parent_peer_sender.send_message(msg_tx(tx_parent_bad_wit))
+
+        # 4. Transactions should not be in mempool.
+        node_mempool = node.getrawmempool()
+        assert tx_parent_bad_wit.rehash() not in node_mempool
+        assert tx_orphan_bad_wit.rehash() not in node_mempool
+
+        # 5. Peer sent a consensus-invalid transaction.
+        parent_peer_sender.wait_for_disconnect()
+
+        # 6. Child-sending peer unpunished for now!
+        child_peer_sender.sync_with_ping()
 
     def run_test(self):
         node = self.nodes[0]
         self.wallet = MiniWallet(node)
         self.wallet_nonsegwit = MiniWallet(node, mode=MiniWalletMode.RAW_P2PK)
         self.generate(self.wallet_nonsegwit, 10)
         self.generate(self.wallet, 20)
 
         fill_mempool(self, node, self.wallet)
 
         self.log.info("Check opportunistic 1p1c logic when parent (txid != wtxid) is received before child")
         self.test_basic_parent_then_child(self.wallet)
 
         self.log.info("Check opportunistic 1p1c logic when parent (txid == wtxid) is received before child")
         self.test_basic_parent_then_child(self.wallet_nonsegwit)
 
         self.log.info("Check opportunistic 1p1c logic when child is received before parent")
         self.test_basic_child_then_parent()
 
         self.log.info("Check opportunistic 1p1c logic when 2 candidate children exist (parent txid != wtxid)")
         self.test_low_and_high_child(self.wallet)
 
         self.log.info("Check opportunistic 1p1c logic when 2 candidate children exist (parent txid == wtxid)")
         self.test_low_and_high_child(self.wallet_nonsegwit)
 
         self.log.info("Check opportunistic 1p1c logic with consensus-invalid orphan causes disconnect")
         self.test_orphan_consensus_failure()
 
+        self.log.info("Check opportunistic 1p1c logic doesn't evaluate package with consensus-invalid parent")
+        self.test_parent_consensus_failure()
+
 
 if __name__ == '__main__':
     PackageRelayTest().main()
```

instagibbs commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,241 @@
+    def test_low_and_high_child(self, wallet):
+        node = self.nodes[0]
+        low_fee_parent = wallet.create_self_transfer(fee_rate=FEERATE_1SAT_VB, confirmed_only=True)
+        low_fee_child = wallet.create_self_transfer(utxo_to_spend=low_fee_parent["new_utxo"], fee_rate=2*FEERATE_1SAT_VB)
```
suggestion:
call this `med_fee_child` and use the mempoolminfee directly since that's a more meaningful value than `2*FEERATE_1SAT_VB` which is below the floating minfee

glozow replied:
done

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,115 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    // package acceptance. Create a random permutation of the indices.
+    std::vector<size_t> tx_indices(cpfp_candidates.size());
+    std::iota(tx_indices.begin(), tx_indices.end(), 0);
+    Shuffle(tx_indices.begin(), tx_indices.end(), m_rng);
```
would it make sense to match peers and (if possible) always pick a child from the orphanage that was sent to us by the same peer that sent us the parent (instead of a random one)?
That way, it wouldn't be possible that a third peer could send us multiple low-fee children that we'd store in the orphanage, in the hope that we pick one of those and reject the package.

instagibbs replied:
I think I made this point in person from the philosophical standpoint that orphanage-churning aside, we probably shouldn't allow peers to "cross-talk" when it comes to package evaluation and possible punishment.

In practice I think an adversary can just churn the orphanage, but still might be a good conceptual framework to adhere to?

glozow replied:
great idea, will change 👍

instagibbs replied:
For some code deduplication you could just have `GetChildrenFromPeer` which takes `std::optional<NodeId>`, returning full `std::vector<std::pair<CTransactionRef, NodeId>>` list if `std::nullopt` or filtered to the specific nodeid otherwise. You can call it filtered once, then if a suitable orphan isn't found, call it again unfiltered.

glozow replied:
I don't think that is much simpler. Also I think this is easier to delete if/when we don't try orphans from different peers in the future.

sr-gi reviewed (commented):
First pass. Reviewed up to 91f4efa420958a93f4620379f8830231f276b23b

glozow reviewed (commented):
Main changes:
- Try same-peer orphanage children (most recent first) before different-peer ones (randomized). Replaced `GetChildren` with `GetChildrenFromSamePeer` and `GetChildrenFromDifferentPeer`.
- Deleted `GetCombinedHash` because it's unused. Also deleted the tests for it
- Changed `GetPackageHash` a bit. Using single-sha256 instead of double-sha256 and using just wtxids instead of serialized vector

glozow commented:
debugging the p2p_opportunistic_1p1c.py failure. I think the wallet one is unrelated.

glozow commented:
Ok ready for review again.

instagibbs commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,319 @@
+        ]]
+        self.supports_cli = False
+
+    def subest_check(self):
```
duplicated `cleanup`, or something?

glozow replied:
Ah forgot to delete, yes

instagibbs commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,319 @@
+        # 5. Peer sent a consensus-invalid transaction.
+        fake_parent_sender.wait_for_disconnect()
+
+        self.log.info("Check opportunistic 1p1c logic with consensus-invalid parent causes disconnect of the correct peer")
```
duplicate log?

sr-gi reviewed (commented):
Second pass (up to the same height: 42859548ab5aebf40da6089b85065f7c204b992a)

Left some additional comments on the updated approach.

glozow commented on src/txorphanage.cpp:
```diff
@@ -241,3 +241,71 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
+    LOCK(m_mutex);
+
+    // First construct vector of iterators to ensure we do not return duplicates of the same tx.
+    std::vector<OrphanMap::iterator> unique_iters;
```
I copied this from the way we do `unique_parents` in orphan parent requests (see the discussion on #19596 about the dynamic memory usage and speed). However I don't mind either way and agree a `std::set` would be simpler, so happy to change if people prefer using a set.

glozow commented on src/test/orphanage_tests.cpp:
```diff
@@ -138,4 +162,154 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
+        std::set<CTransactionRef> expected_parent1_children{child_p1n0, child_p1n0_p2n0, child_p1n0_p1n1};
+        std::set<CTransactionRef> expected_parent2_children{child_p2n1, child_p1n0_p2n0};
+
+        // Check contents and ensure transactions are returned in order of recency.
```
Ah, forgot to remove the comment. We need to make time parameterizable across `TxOrphanage` members to test order, so not testing / saving for followup.

instagibbs reviewed (commented):
reviewed through 7d220c6a5c0e0c5e8cfe79ebd2eae6e845d1d983

tested and confirmed fuzz coverage is hitting meaningful `GetChildrenFrom*` results

continuing longer range testing

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,136 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    // These children should be sorted from most newest to oldest. In the case of children that
+    // replace each other, this helps us accept the highest feerate (most recent) one most
+    // efficiently.
+    for (const auto& child : cpfp_candidates_same_peer) {
```
Just an observation: I guess the more common case will likely be multiple children of the same parent transaction (rather than conflicting children).  Still, we have no idea which child will have the best chance of successfully bumping it, so this ordering seems as good as any.

glozow replied:
Edited the comment to say this is a probably uncommon case

sdaftuar reviewed (approved):
Code review ACK (apart from the tests, which I only skimmed).  Will test...

glozow commented on src/test/orphanage_tests.cpp:
```diff
@@ -138,4 +162,157 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
+            BOOST_CHECK(expected_parent2_children.contains(tx));
+        }
+        // The peer must match
+        BOOST_CHECK(orphanage.GetChildrenFromSamePeer(parent1, node2).empty());
```
Done. Also refactored orphanage_tests to be more readable.

glozow commented on src/net_processing.cpp:
```diff
@@ -4354,7 +4354,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
-            ProcessValidTx(pfrom.GetId(), ptx, result.m_replaced_transactions.value());
+            Assume(result.m_replaced_transactions.has_value());
```
I definitely it's awkward and doesn't have much benefit. I'll leave the assume/handling here, and maybe in a followup we can change it to non-optional.

glozow replied:
Done in #30012

glozow commented on src/net_processing.cpp:
```diff
@@ -4432,7 +4586,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-                    if (!AlreadyHaveTx(gtxid, /*include_reconsiderable=*/true)) AddTxAnnouncement(pfrom, gtxid, current_time);
+                    // Exclude m_recent_rejects_reconsiderable: the missing parent may have been
+                    // previously rejected for being too low feerate. This orphan might CPFP it.
+                    if (!AlreadyHaveTx(gtxid, /*include_reconsiderable=*/false)) AddTxAnnouncement(pfrom, gtxid, current_time);
```
Good point, I don't think this is too complex so I've added it.

While working on this, I remembered/realized that if there are any other unconfirmed parents at all, we will reject it because the package must be child-with-unconfirmed-parents, enforced here:
https://github.com/bitcoin/bitcoin/blob/256e1703197fdddd78bc6d659431cd0fc3b63cde/src/validation.cpp#L1558-L1564

IIRC this was the way to check that a package was "2 generations only". But I don't know how useful this is, and it's quite annoying here, so perhaps we should consider getting rid of this restriction in a followup... Then we can accept packages where the child has parents already in mempool.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,140 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+        const Package package_1p1c{ptx, tx_orphan};
+        const std::vector<NodeId> senders{nodeid, orphan_sender};
+        const auto package_result{ProcessNewPackage(m_chainman.ActiveChainstate(), m_mempool, package_1p1c, /*test_accept=*/false, /*client_maxfeerate=*/std::nullopt)};
+        LogDebug(BCLog::TXPACKAGES, "package evaluation for parent %s (wtxid=%s) + child %s (wtxid=%s) in orphanage: %s\n",
```
FWIW: I added logging here to see what number of parents and given by different peers than the orphan. After a day of running it's 0 out of 162. 

Is there a reason to think this should be a common pattern?

sdaftuar replied:
If a parent+child are relayed back to back across the network, then I would expect there to be times when we download the transactions out of order (eg because we send a request for the parent to a different peer than the request for the child).

instagibbs replied:
I guess it'll likely be something like:

1) peer A sends child, child is put in orphanage and parent tx request queued(but not yet sent)
2) peer B has slightly out of date feefilter for your node, sends INV for parent

glozow replied:
fwiw it's definitely not common, but I do see them occasionally and have a handful of acceptances of packages from 2 different peers

glozow replied:
> though I'm not convinced this really will make a difference

sdaftuar replied:
> I guess it'll likely be something like:
> 
> 1. peer A sends child, child is put in orphanage and parent tx request queued(but not yet sent)
> 2. peer B has slightly out of date feefilter for your node, sends INV for parent

What I'd expect to happen is that we get inv's for parent+child from both Peer A and Peer B, and we happen to request the parent from A and the child from B, but the child arrives first -- it's put in the orphanage, and since a request for the parent is already in flight to A, we don't send another request for the parent to B.

instagibbs replied:
> Is the suggestion to drop this and only try same-peer packages?

If this is hitting actual usage, no, because it's coded and I actually reviewed it, and tests seem to cover it. I'm unsure if in BIP331-like world it would be used since it would be receiver-driven, but that might be thinking too far ahead.

> (In this scenario I guess I'm assuming no feefilter, or that the feefilter value is slightly stale.)

glozow replied:
Ok, marking this as resolved

instagibbs reviewed (commented):
reviewed through 55b1280c52af81aa6ea0860799fa16da49f51447

Looks good, have a suggestion for a test catching the `Assume()` issue(which does actually get hit in mainnet pretty fast with debug on)

glozow commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,348 @@
+        assert "getdata" not in peer_sender.last_message
+
+
+    def run_test(self):
```
Added a similar test

glozow commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,348 @@
+
+        # 3. Node should not request any parents, as it should recognize that it will not accept
+        # multi-parent-1-child packages.
+        node.bumpmocktime(60)
```
Added

instagibbs reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/28970/commits/30c9e6bc4e9f8b8b606e55435dc3f743cb2dd670

Changes are test only, and new subtest causes a debug-build crash when the `PCKG_POLICY` change is reverted.

theStack reviewed (commented):
Another round through (modulo the fuzz commit), the code looks correct to me, just left a few nits below. Planning to do another review tomorrow with focus on the main commit bb3bf0a39c036fbd94b99ed7002db7d9bdf0a1d6, especially going through the `AlreadyHaveTx` call-sites and their `include_reconsiderable` values to check once more.

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,140 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+            }
+        }
+        package_iter++;
+        // There should be a sender for each tx, but fall back to the last one otherwise.
+        if (Assume(senders_iter != senders.rend())) senders_iter++;
```
I think we'd want to make sure that even if we had a bad `senders` vector we would still call `ProcessValidTx` / `ProcessInvalidTx` (seems kind of bad not to). Maybe a better approach is to use -1 when we don't know what it is - I do realize we're not handling the empty vector case.

(Yeah maybe this is moot as the 1 caller should always make `senders` properly...)

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -40,6 +43,92 @@ inline CTransactionRef create_placeholder_tx(size_t num_inputs, size_t num_outpu
+    uint256 hash_if_use_txid = (HashWriter() << txid_2 << txid_1 << txid_3).GetSHA256();
+    BOOST_CHECK(hash_if_use_txid != calculated_hash_123);
+
+    uint256 hash_if_use_int_order = (HashWriter() << wtxid_2 << wtxid_1 << txid_3).GetSHA256();
```
Yeah that's a typo :+1:

sr-gi reviewed (commented):
Third pass, reviewing the changes of the first two, plus going all the way to 30c9e6bc4e9f8b8b606e55435dc3f743cb2dd670

I think the code looks good overall, but I have some questions regarding some corner cases that I'm not completely sure about (see inline).

Finally, something I've noticed: Given a `1p1c` whose parent we've seen already and we just received the child, the parent will be tried on its own twice before being considered as a package. This is because we will try it first when receiving it, another time when submitting it as a package, by splitting the package, and a third time alongside the child. My understanding of this is that we try to submit the smallest package that cannot make it on its own (I'm assuming to prevent some pathological cases). This makes sense in the general case, but not so much in the `1p1c`. I'm not sure having additional complexity to treat this specific case would be worth it though.

sr-gi commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -0,0 +1,392 @@
+        # 6. Child-sending should not have been punished and the orphan should remain in orphanage.
+        # It can send the "real" parent transaction, and the package is accepted.
+        package_sender.sync_with_ping()
+        package_sender.send_and_ping(msg_tx(low_fee_parent["tx"]))
```
> this is from `test_orphan_consensus_failure` causing the exact same parent to be added to `m_recent_rejects_reconsiderable`
> 
> Can we somehow cause chaintip updates during cleanup, without dropping minfee too much? What if cleanup detects that it went down too much, then calls `fill_mempool` again?

lol, I missed this 🤣

glozow replied:
> Can we somehow cause chaintip updates during cleanup, without dropping minfee too much? What if cleanup detects that it went down too much, then calls fill_mempool again?

Hm, `fill_mempool` isn't really built for this (though we can change that). I'll try to see if we can make sure we don't reuse utxos.

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,140 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+                tx_orphan = cpfp_candidates_different_peer.at(index).first;
+                orphan_sender = cpfp_candidates_different_peer.at(index).second;
+                break;
+            }
```
done

glozow commented:
Addressed @theStack and @sr-gi comments. Thanks!

instagibbs commented on src/net_processing.cpp:
```diff
@@ -596,6 +596,45 @@ class PeerManagerImpl final : public PeerManager
+    void ProcessPackageResult(const Package& package, const PackageMempoolAcceptResult& package_result, const std::vector<NodeId>& senders)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, g_msgproc_mutex, cs_main);
+
+    /** A package to validate  */
```
redundant comment

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3198,6 +3237,117 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
     }
 }
 
+void PeerManagerImpl::ProcessPackageResult(const Package& package, const PackageMempoolAcceptResult& package_result, const std::vector<NodeId>& senders)
```
now that we have `PackageToValidate`, just directly pass here?

glozow replied:
Done in #30012

instagibbs commented on src/net_processing.cpp:
```diff
@@ -596,6 +596,45 @@ class PeerManagerImpl final : public PeerManager
+            m_senders {parent_sender, child_sender}
+        {}
+
+        std::string ToString() const {
```
useful log message for stats :+1:

instagibbs reviewed (approved):
reACK e518a8bf8abf3d7b83c9013f56d0dca18ae04d6f

non-blocking comments

reviewed via `git range-diff master 30c9e6b e518a8bf8abf3d7b83c9013f56d0dca18ae04d6f`

sr-gi commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,140 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    const auto& parent_txid{ptx->GetHash()};
+    const auto& parent_wtxid{ptx->GetWitnessHash()};
+
+    CTransactionRef tx_orphan{nullptr};
```
Much cleaner now

glozow commented on src/net_processing.cpp:
```diff
@@ -3195,6 +3210,140 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+    AssertLockHeld(g_msgproc_mutex);
+    AssertLockHeld(cs_main);
+
+    if (package_result.m_state.IsInvalid()) {
+        m_recent_rejects_reconsiderable.insert(GetPackageHash(package));
+    }
```
Same question as https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1566378431

I like that this doesn't add extra burden to `m_recent_rejects` (which is probably the busier of the 2 filters even though they have the same size). It makes more sense if you think of each tx within the package as reconsiderable, even though the package isn't?

sr-gi replied:
I think, design-wise, it's a bit counter-intuitive, but I do agree that not overloading `m_recent_rejects` is potentially better

theStack reviewed (approved):
Code-review ACK e518a8bf8abf3d7b83c9013f56d0dca18ae04d6f :package:

glozow commented:
Since there are a few ACKs now, listing followups. I plan to open a PR for the first two immediately:
- redundant comment + pass `PackageToValidate` directly https://github.com/bitcoin/bitcoin/pull/28970#pullrequestreview-2025023787
- make `MempoolAcceptResult::m_replaced_transactions` non-optional https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1568781077
- (already in #29974) fix quirks in fuzz/txorphan.cpp https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1576401327
- consider dropping requirement that all unconfirmed parents be present, allow package with other parent in mempool https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1576145432. This may conflict with #28984, so will hold off for now.
- cache rejection of orphan if parents are rejected https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1511279226

glozow commented on src/net_processing.cpp:
```diff
@@ -4371,10 +4582,23 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+                } else if (m_recent_rejects_reconsiderable.contains(parent_txid) && !m_mempool.exists(GenTxid::Txid(parent_txid))) {
+                    // More than 1 parent in m_recent_rejects_reconsiderable: 1p1c will not be
+                    // sufficient to accept this package, so just give up here.
+                    if (rejected_parent_reconsiderable.has_value()) {
```
Ah, just found a mistake. I think this can go in a followup though.
```suggestion
                    if (rejected_parent_reconsiderable.has_value() && *rejected_parent_reconsiderable != parent_txid.ToUint256()) {
```

sdaftuar replied:
Hmm, we're looping over unique parents already, so I think this code should be fine?  In which case is there a bug?

instagibbs replied:
they're unique txids, which scenarios will this fix?

glozow replied:
Ah you're right, nevermind!

dergoegge commented on src/txorphanage.cpp:
```diff
@@ -241,3 +241,77 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
+        }
+    }
+
+    // Erase duplicates
+    std::sort(iters.begin(), iters.end(), IteratorComparator());
+    iters.erase(std::unique(iters.begin(), iters.end()), iters.end());
```
Why is this sort comparison different from the one in `GetChildrenFromSamePeer`?

These functions are almost identical as well. Perhaps a more generic ```std::vector<std::pair<CTransactionRef, NodeId>> GetChildren(const CTransactionRef&)``` that returns all children and their senders would be less code?

mzumsande replied:
I agree, would be nice to avoid this code duplication.

instagibbs replied:
made a similar point at https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1571063644 fwiw (with a response attached)

glozow replied:
The idea is to return `GetChildrenFromSamePeer` in recency order. This one could be returned that way as well, as the goal is to have a random order (and we randomize the indices), but all I wanted to do here was deduplicate.

As for the 2 functions having duplicate code, for a longer explanation: I'm hoping we can remove 1p1cs with different providers by tracking all orphan announcers, which would mean deleting `GetChildrenFromDifferentPeer`. I think it'll be easier to delete the function + unit tests wholesale if it's a different function. But it seems I'm in the minority here, and apologies for resolving the first comment about this. Will avoid this in the future.

dergoegge reviewed (commented):
light Code review ACK e518a8bf8abf3d7b83c9013f56d0dca18ae04d6f

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3097,7 +3133,14 @@ void PeerManagerImpl::ProcessInvalidTx(NodeId nodeid, const CTransactionRef& ptx
+            // If the result is TX_RECONSIDERABLE, add it to m_recent_rejects_reconsiderable
+            // because we should not download or submit this transaction by itself again, but may
+            // submit it as part of a package later.
+            m_recent_rejects_reconsiderable.insert(ptx->GetWitnessHash().ToUint256());
```
nit: I think that 6c51e1d7d021ed6523107a6db87a865aaa8fc4c9 changes behavior such that txns which are now in `m_recent_rejects_reconsiderable` instead of `m_recent_rejects` will be ignored when calling `m_recent_rejects.contains(...)`. That is probably an undesired change of behavior - however it's just for 1 commit :man_shrugging:

achow101 reviewed (commented):
ACK e518a8bf8abf3d7b83c9013f56d0dca18ae04d6f

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3198,6 +3237,117 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
+
+    // No package results to look through for PCKG_POLICY or PCKG_MEMPOOL_ERROR
+    if (package_result.m_state.GetResult() == PackageValidationResult::PCKG_POLICY ||
+        package_result.m_state.GetResult() == PackageValidationResult::PCKG_MEMPOOL_ERROR) return;
```
I wonder if there are situations in which we fail on a package level and therefore return here and never call `ProcessInvalidTx()` to remove the child from the orphanage. Would that be possible, for example, with v3 transactions that violate `PackageV3Checks()`, and is there something we could do against it? (or should, maybe it's not that terrible?).

glozow replied:
`PCKG_POLICY` doesn't populate `m_tx_results` so we wouldn't have results for those transactions - it's meant for situations in which the package is invalid, but the individual transactions could be if submitted in a different context. However, going over the failures that are marked as `PCKG_POLICY`, I agree there's a few failures that are classified as `PCKG_POLICY` but are actually more definitively invalid, and we should bubble that up to p2p.

Having both mempool + package v3 parents, v2-spending-v3, and v3-spending-v2 are failures that can't change in another context, so we can delete the tx from orphanage.

We shouldn't do this for `package-mempool-limits` and `package-not-child-with-unconfirmed-parents`, since there could be a parent with a smaller witness that makes this child acceptable.

The 2 changes we'd make are:
- update validation-side things to return a `MempoolAcceptResult` for the results where we should mark transactions as invalid. Perhaps we can reclassify these failures as `PCKG_TX` failures, or allow `PCKG_POLICY` failures to populate `m_tx_results`. cc @instagibbs ?
- update the p2p logic here to continue when it's `PCKG_POLICY` and call `ProcessInvalidTx` for the `MempoolAcceptResult`s we find.

instagibbs replied:
> PCKG_POLICY doesn't populate m_tx_results so we wouldn't have results for those transactions

From my reading of the code, `package-mempool-limits` populates `m_tx_results`? We also allow those cases in our fuzz target(which I'm guessing is there because I hit it!).

We could probably be more aggressive about filling out results when possible and passing those results along, sounds like future work for now?

glozow replied:
Then maybe we should just do this?

> Update the p2p logic here to continue when it's PCKG_POLICY and call ProcessInvalidTx for the MempoolAcceptResults we find.

(and remove the `Assume` that they always exist)

glozow replied:
^I added this to #30012

mzumsande reviewed (commented):
Light ACK e518a8bf8abf3d7b83c9013f56d0dca18ae04d6f
I didn't review the functional tests in much detail, but the p2p code looks good to me.

glozow commented on src/policy/packages.cpp:
```diff
@@ -147,3 +147,21 @@ bool IsChildWithParentsTree(const Package& package)
+        [](const auto& tx){ return tx->GetWitnessHash(); });
+
+    // Sort in ascending order
+    std::sort(wtxids_copy.begin(), wtxids_copy.end(), [](const auto& lhs, const auto& rhs) { return lhs.GetHex() < rhs.GetHex(); });
```
> Is there a reason to convert the wtxids into hex strings for this comparison? That seems kind of expensive, especially when Wtxid already has an operator<.

> I surmise that this has to do with the "concatenated in lexicographical order (treating the wtxids as little endian encoded uint256, smallest to largest)" and BIP 331 defines the same thing, but it's not clear to me why it must be like that (and couldn't also be changed in the BIP).

The code was originally using the operator< defined on `uint256`, and the opposite switch happened :sweat_smile: A few people commented on lexicographical being underspecified or different from how it's usually defined (e.g. https://github.com/bitcoin/bitcoin/pull/28970#discussion_r1566025027). Mostly I felt like reverse byte order was much more intuitive, so I changed it here and in the BIP. Yes this doesn't have to match the BIP since it's only used internally, but I figured I might as well write something that's reusable.

Thanks for the `std::lexicographical_compare` suggestion, I'll add that in a followup.

glozow replied:
Done in #30012
